--- qemu-system-x86_64.js 2025-09-04 22:29:54.371974611 +0200
+++ qemu-system-x86_64.js	2025-09-04 22:32:29.154423162 +0200
@@ -1,6 +1,5 @@
 // This code implements the `-sMODULARIZE` settings by taking the generated
 // JS program code (INNER_JS_CODE) and wrapping it in a factory function.
-
 // When targetting node and ES6 we use `await import ..` in the generated code
 // so the outer function needs to be marked as async.
 async function Module(moduleArg = {}) {
@@ -1066,15 +1065,20 @@
 }

 var wasmTableMirror = [];
+var wasmTableIsAsync = [];

 /** @type {WebAssembly.Table} */ var wasmTable;

-var getWasmTableEntry = funcPtr => {
+var getWasmTableEntry = (funcPtr, forceAsync=false) => {
   var func = wasmTableMirror[funcPtr];
+  if (forceAsync && func && !wasmTableIsAsync[funcPtr]) {
+    func = null;
+  }
   if (!func) {
     /** @suppress {checkTypes} */ wasmTableMirror[funcPtr] = func = wasmTable.get(funcPtr);
-    if (Asyncify.isAsyncExport(func)) {
+    if (forceAsync || Asyncify.isAsyncExport(func)) {
       wasmTableMirror[funcPtr] = func = Asyncify.makeAsyncFunction(func);
+      wasmTableIsAsync[funcPtr] = true;
     }
   }
   return func;
@@ -1101,7 +1105,7 @@
   // *ThreadMain(void *arg) form, or try linking with the Emscripten linker
   // flag -sEMULATE_FUNCTION_POINTER_CASTS to add in emulation for this x86
   // ABI extension.
-  var result = WebAssembly.promising(getWasmTableEntry(ptr))(arg);
+  var result = getWasmTableEntry(ptr, true)(arg);
   function finish(result) {
     if (keepRuntimeAlive()) {
       EXITSTATUS = result;
@@ -8331,8 +8335,7 @@
           Asyncify.asyncExports.add(original);
           original = Asyncify.makeAsyncFunction(original);
         }
-        var wrapper = Asyncify.instrumentFunction(original);
-        ret[x] = wrapper;
+        ret[x] = original;
       } else {
         ret[x] = original;
       }
@@ -8383,12 +8386,10 @@
       Fibers.resolvers[promiseId]();
     } else {
       var entryPoint = HEAPU32[(((newFiber) + (12)) >>> 2) >>> 0];
-      // when it suspends we'll write promiseid, so not sure if this is also needed
-      HEAPU32[(((newFiber) + (12)) >>> 2) >>> 0] = 0;
       var userData = HEAPU32[(((newFiber) + (16)) >>> 2) >>> 0];
       // KOKO: this calls .promising() each time
       // while only once per coroutine, this doesn't use cache;
-      WebAssembly.promising(getWasmTableEntry(entryPoint))(userData);
+      getWasmTableEntry(entryPoint, true)(userData);
     }
   }
 };
@@ -11658,9 +11659,9 @@

 var dynCall = (sig, ptr, args = [], promising = false) => {
   var func = getWasmTableEntry(ptr);
-  if (promising) {
-    func = WebAssembly.promising(func);
-  }
+  //if (promising) {
+  //  func = WebAssembly.promising(func);
+  //}
   var rtn = func(...args);
   function convert(rtn) {
     return sig[0] == "p" ? rtn >>> 0 : rtn;
@@ -12219,7 +12220,7 @@
   return [ type_ptr, type_id ];
 }

-function ffi_call_js(cif, fn, rvalue, avalue) {
+async function ffi_call_js(cif, fn, rvalue, avalue) {
   var abi = HEAPU32[(cif >> 2) + 0 >>> 0];
   var nargs = HEAPU32[(cif >> 2) + 1 >>> 0];
   var nfixedargs = HEAPU32[(cif >> 2) + 6 >>> 0];
@@ -12375,7 +12376,7 @@
   stackRestore(cur_stack_ptr);
   stackAlloc(0);
   0;
-  var result = getWasmTableEntry(fn).apply(null, args);
+  var result = await ((getWasmTableEntry(fn, true)).apply(null, args));
   stackRestore(orig_stack_ptr);
   if (ret_by_arg) {
     return;
